File: src/server/services/scanner.ts
import { readdir, stat } from "fs/promises";
import { join } from "path";
import { homedir } from "os";

const PROJECTS_DIR = join(homedir(), ".claude", "projects");

export interface ScannedProject {
  id: string;
  dir: string;
  sessionFiles: { id: string; path: string; mtime: Date; size: number }[];
}

export async function scanProjects(): Promise<ScannedProject[]> {
  const entries = await readdir(PROJECTS_DIR, { withFileTypes: true });
  const projects: ScannedProject[] = [];

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const projectDir = join(PROJECTS_DIR, entry.name);
    const files = await readdir(projectDir, { withFileTypes: true });

    const sessionFiles: ScannedProject["sessionFiles"] = [];
    for (const f of files) {
      if (!f.name.endsWith(".jsonl") || f.isDirectory()) continue;
      const filePath = join(projectDir, f.name);
      const s = await stat(filePath);
      sessionFiles.push({
        id: f.name.replace(".jsonl", ""),
        path: filePath,
        mtime: s.mtime,
        size: s.size,
      });
    }

    if (sessionFiles.length > 0) {
      projects.push({
        id: entry.name,
        dir: projectDir,
        sessionFiles,
      });
    }
  }

  return projects;
}

export function getProjectsDir(): string {
  return PROJECTS_DIR;
}

File: cli/src/scanner.rs
use anyhow::Result;
use std::fs;
use std::path::PathBuf;

use crate::models::{ScannedProject, SessionFile};

pub fn get_projects_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("~"))
        .join(".claude")
        .join("projects")
}

pub fn scan_claude_projects() -> Result<Vec<ScannedProject>> {
    let projects_dir = get_projects_dir();

    if !projects_dir.exists() {
        return Ok(vec![]);
    }

    let mut projects = Vec::new();

    let entries = fs::read_dir(&projects_dir)?;
    for entry in entries {
        let entry = entry?;
        if !entry.file_type()?.is_dir() {
            continue;
        }

        let project_dir = entry.path();
        let project_id = entry.file_name().to_string_lossy().to_string();

        let mut session_files = Vec::new();
        if let Ok(files) = fs::read_dir(&project_dir) {
            for f in files {
                let f = f?;
                let fname = f.file_name().to_string_lossy().to_string();
                if !fname.ends_with(".jsonl") || f.file_type()?.is_dir() {
                    continue;
                }
                let meta = f.metadata()?;
                session_files.push(SessionFile {
                    id: fname.trim_end_matches(".jsonl").to_string(),
                    path: f.path().to_string_lossy().to_string(),
                    size: meta.len(),
                });
            }
        }

        if !session_files.is_empty() {
            projects.push(ScannedProject {
                id: project_id,
                dir: project_dir.to_string_lossy().to_string(),
                session_files,
            });
        }
    }

    Ok(projects)
}
